File: Java/2_Stack/Stack.java
import java.util.Scanner;

public class Stack {

    private int top;
    private int maxSize;
    private int[] stack;

    Stack(int maxSize){
        this.maxSize = maxSize;
        this.top = -1;
        this.stack = new int[this.maxSize]; 
    }

    public void push(Scanner sc){

        if (this.top == this.maxSize-1){
            System.out.println("Stack OverFlow\n");
            return;
        }

        System.out.print("Enter The Value to Push: ");
        int value = sc.nextInt();
        stack[++this.top] = value;
        System.out.println(value+" pushed to stack\n");

    }

    public void pop(){
        if(this.top == -1){
            System.out.println("Stack UnderFlow\n");
            return;
        }
        System.out.println(stack[this.top--]+" Poped\n");
        return;
    }

    public void display(){

        if(this.top == -1){
            System.out.println("Stack is Empty");
            return;
        }

        for(int i = 0; i<=this.top; i++){
            System.out.print(stack[i]+" ");
        }
        System.out.println("\n");

    }

    public void peek(){
        if (this.top == -1){
            System.out.println("Stack UnderFlow ");
            return;
        }

        System.out.println(stack[this.top]+" Top Element");
        return;
    }
    
}

File: DSA/01 - 3/stack.c
#include <stdio.h>
#include <stdlib.h>
#define MAX 5

int stack[MAX], top = -1;

void push() {
    if (top == MAX - 1) {
        printf("Stack Overflow\n");
        return;
    }
    int item;
    printf("Enter element: ");
    scanf("%d", &item);
    stack[++top] = item;
}

void pop() {
    if (top == -1) {
        printf("Stack Underflow\n");
        return;
    }
    printf("Popped: %d\n", stack[top--]);
}

void display() {
    if (top == -1) {
        printf("Stack is Empty\n");
        return;
    }
    printf("Stack elements:\n");
    for (int i = top; i >= 0; i--)
        printf("%d\n", stack[i]);
}

void palindrome() {
    if (top == -1) {
        printf("Stack is empty\n");
        return;
    }
    for (int i = 0, j = top; i < j; i++, j--) {
        if (stack[i] != stack[j]) {
            printf("Not Palindrome\n");
            return;
        }
    }
    printf("Palindrome\n");
}

int main() {
    int choice;
    while (1) {
        printf("\n1.Push\n2.Pop\n3.Palindrome\n4.Display\n5.Exit\nChoice: ");
        scanf("%d", &choice);
        switch (choice) {
            case 1: push(); break;
            case 2: pop(); break;
            case 3: palindrome(); break;
            case 4: display(); break;
            case 5: exit(0);
            default: printf("Invalid choice\n");
        }
    }
}
